[Async Refactoring] Better handle bool flag parameter bindings

hamishknight opened the pull request:
If we have a known bool flag parameter, drop it from the success parameters being bound in a refactored await call, as the async variant drops it. And in the fallback case, add a binding for it in the success and error block.

rdar://81896460

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7109,42 +7199,23 @@ class AsyncConverter : private SourceEntityWalker {
+                                       DiagEngine, CallbackBody);
     }
 
+    SmallVector<const ParamDecl *, 4> Scratch;
+    auto SuccessBindings = CallbackParams.getSuccessParamsToBind(Scratch);
```
Given this is the only use I'd prefer just returning the SmallVector from `getSuccessParamsToBind`.

And not sure if this part is worth it or not but...
  - Could name `SuccessBindings` just `Bindings`
  - Add an `ArrayRef` `SuccessBindings` of `Bindings`
  - Remove `AllBindings` and just add `ErrParam` to `Bindings` in the if

hamishknight replied:
Hmm, returning a SmallVector directly would mean the callee would have to specify the inline buffer size, which feels odd – maybe we could return a `llvm::TinyPtrVector` or `std::vector`?

Regarding mutating a single `Bindings` vector directly, I tend to prefer reducing mutability where I can, but this is all pretty localised so I don't feel too strongly about it

hamishknight replied:
I ended up generalising this logic to not need the mutation. I did switch over using `llvm::TinyPtrVector` though, I don't really feel too strongly about the return type tho

bnbarham replied:
`TinyPtrVector` is optimised for 0/1 right? I guess that is probably almost always true for the success/error. I'm fine with specifying the inline buffer size in the return type, but I don't *really* mind. Personally I'd still use `SmallVector<Thing, 0>` over a vector, even if I was returning it.

hamishknight replied:
Yeah `TinyPtrVector` is optimized for the single element (and empty) case, which as you say is fairly likely for the success/error params. Out of curiosity, what benefits would `SmallVector<Thing, 0>` have over a regular vector?

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7157,22 +7228,29 @@ class AsyncConverter : private SourceEntityWalker {
       return;
     }
 
+    auto *ErrOrResultParam = ErrParam;
```
Better name, thanks :P

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7206,34 +7284,34 @@ class AsyncConverter : private SourceEntityWalker {
-                                  /*Success=*/true);
+    prepareNames(Blocks.SuccessBlock, SuccessBindings, InlinePatterns);
+    preparePlaceholdersAndUnwraps(HandlerDesc, CallbackParams,
+                                  PlaceholderMode::SUCCESS_BLOCK);
```
🙇

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7513,12 +7591,31 @@ class AsyncConverter : private SourceEntityWalker {
+      if (!Params.hasBinding(Param))
+        Placeholders.insert(Param);
+    }
+    // For the fallback case, no other params need placeholdering, as they are
+    // all freely accessible in the fallback case.
+    if (Mode == PlaceholderMode::FALLBACK)
+      return;
```
I mention this in the tests below, but IMO we shouldn't placeholder the bool param either. As mentioned there though, happy to do that later.

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4963,6 +4963,117 @@ class ClassifiedBlock {
+      : HandlerDesc(HandlerDesc),
+        AllParams(Closure->getParameters()->getArray()) {
+    assert(AllParams.size() == HandlerDesc.params().size());
+    assert(!(HandlerDesc.Type == HandlerType::RESULT && AllParams.size() != 1));
```
Heh that's how I had it initially but thought this was marginally more readable. I don't really feel too strongly either way

hamishknight replied:
Changed it over

hamishknight commented on test/refactoring/ConvertAsync/convert_bool.swift:
```diff
@@ -461,4 +459,21 @@ func testConvertBool() async throws {
   // OBJC-BOOL-WITH-ERR2-NEXT:   print("neato")
   // OBJC-BOOL-WITH-ERR2-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 -I %S/Inputs -I %t -target %target-triple %clang-importer-sdk-nosource | %FileCheck -check-prefix=OBJC-BOOL-WITH-ERR-FALLBACK %s
```
Good point regarding setting `success` in the different branches! I can look into doing that, should be reasonably straightforward.

The reason why the fallback is being used is because we have an unhandled condition mixed with a classified condition, and we have an `else` block, so we don't know what to do with it, as the `else` block could be run under success or failure.

hamishknight replied:
Added logic to bind the `success` variable, how does it look?

bnbarham replied:
LGTM, thanks :)

hamishknight committed:
[Async Refactoring] Add ClosureCallbackParams

Refactor some closure parameter handling logic
into a new ClosureCallbackParams type, which lets
us clean up some of the parameters passed to
CallbackClassifier, and better distinguish between
an error param and a Result param.

hamishknight committed:
[Async Refactoring] Drop bool flag parameter binding

If we have a known bool flag parameter, drop it
from the success parameters being bound in a
refactored await call, as the async variant drops
it.

rdar://81896460

hamishknight committed:
[Async Refactoring] Placeholder dropped params in fallback case

Make sure to convert any references to parameters
that get dropped in the fallback case to
placeholders.

hamishknight committed:
[Async Refactoring] Bind known bool param in fallback

Generalize the logic to handle different BlockKinds,
and add binding logic that lets us assign `true` or
`false` to the given bool success param in the
fallback case.

hamishknight commented:
@swift-ci please test
